In [1]:
# Question 1
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [2]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [3]:
# Question 2
#2.1
# Get the shape of the DataFrame
rows, columns = df.shape

# Rows
print(df.shape[0])
#columns
print(df.shape[1])

#2.2
# An observation would be a single row, which records the data of a single villager
# A variable would be one specific data recorded within an observation, such as id or name

391
11


In [4]:
#Question 3
# Summary statistics for numerical columns
print(df.describe())

# Summary statistics for all columns (including non-numerical)
print(df.describe(include='all'))



            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000
             row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
ma

In [5]:
#Print count of each unique value using comprehension, not using .describe() as no numerical values
summary = []
for x in df.columns:
    summary.append(df[x].value_counts())
print(summary)

[row_n
2      1
303    1
331    1
330    1
328    1
      ..
151    1
150    1
149    1
148    1
483    1
Name: count, Length: 391, dtype: int64, id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64, name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64, gender
male      204
female    187
Name: count, dtype: int64, species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey    

In [6]:
# Question 4
titanicurl = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ti = pd.read_csv(titanicurl)
ti

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [7]:
# Rows
print(ti.shape[0])
#columns
print(ti.shape[1])

891
15


In [8]:
print(ti.describe())

         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200


In [9]:
# (a) There is discrepancies between size of dataset given by ti.shape() and ti.describe() becuase .describe() only analyses columns with numeric values
# which is why only survived, pclass, age, sibsp, parch, fare columns are included
# (b) the "count" column has discrepancy because it only counts non-null values
# can be seen in age column, where there are only 714 rows instead of 891

In [10]:
ti.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [11]:
# Question 5
# Didn't use chatbot
# the attribute is data value that belongs to an instance of an object, in this case df
# to access it, you don't need a method, it can just be accessed unless it's private 
# df.shape is simply the values which contain the rows and columns of the data 
# method is a function, which means that it is a function defined by the class, and when called, it will use the data of the object to perform some kind of operation
# in this case, df.describe() provides statistcal analysis of data

In [12]:
# Question 6
# Count is the number of non missing values in the column
# Mean is the average of the sum of values in the colum
# std is standard deviation of the values in the column, and it measures how much the variables are spread from the mean, and can be used to calculate for outliers
# min is the smallest value in the column
# 25% is Q1, which means 25% of data points is in it
# 50% is second quartile, which means half of data is below and other half is above
# 75% is third quartile, 25% data above, 75% data below
# max is the largest value in a column

In [13]:
# Question 7
# 1
# df.dropna() would only remove the rows where there are missing values, for example ti.dropna(subset=['age']) would only remove rows where there are missing values, which doesn't cause other columns to be delete
# del df['col'] deltes the whole column where there might be missing data, which would be undesirable if only a few values were missing
# in the ti example, we would only want to use dropna() because there are only about 170 missing values, with a lot of data remaining, which we want to use

# 2 
# we would use del df['col'] if the column only had missing values or had very little data, making it unable to be used
#in this case, we would rather delete the whole column instead

# 3
# using del df['col'] and ti.dropna() could be important to remove all rows and columns with missing data, in the case you want your statistcal analysis to be more accurate and without missing values
# furthermore, you would remove all redundant columns and rows, which would improve the efficiency of analysis


# 4
# using ti
# Before: 177 missing in age while 688 missing in deck
# will remove only rows for age, as there are still many values available to use
# the majority of data is missing for deck, making it redundant, so the entire column will be removed
ti.isna().sum()


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [14]:
ti.dropna(subset=['age'])

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
885  woman       False  NaN   Queenstown    no  False  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[714 rows x 15 columns]

In [15]:
del ti['deck']


In [16]:
ti
# after the removal

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
888  woman       False  Southampton    no  False  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[891 rows x 14 columns]

In [17]:
# Question 8
# 8.1
# df.groupby("col1")["col2"].describe()
# groupby("col1") groups the unique values in col1 column, ["col2"].describe() will get summary statistics for col2 for each unique value in col1
# For example, using df
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [18]:
df.groupby("gender")["species"].describe()

count unique   top freq
gender                        
female   187     33   cat   14
male     204     34  frog   13

In [19]:
# 8.2
# df.describe() shows how many non missing value for all columns in the dataframe
# df.groupby("col1")["col2"].describe() shows the count of non missing values within each group defined in col1
# for example in df.groupby("gender")["species"].describe() the count would display the count of non missing values in species column, which is sorted by each group from col1, which is male or female


In [20]:
# 8.3A
df
# shows NameError: name 'df' is not defined
# Chatgpt and google was around the same in terms of trouble shooting
# both gave the response that df was not defined, and it needed to be defined, and to check the area where it was defined

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [23]:
# 8.3B
# missing s in villagers
error = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villager.csv"
er = pd.read_csv(error)
# shows HTTPError
# Google and Chatgpt showed the same solutions



HTTPError: HTTP Error 404: Not Found

In [24]:
error = "villager.csv"
er = pd.read_csv(error)
# file not found error
# Google and chatgpt around the same, both said solution was to check the file name and whether path was correct

FileNotFoundError: [Errno 2] No such file or directory: 'villager.csv'

In [25]:
# 8.4C
DF.groupby("col1")
# NameError: name 'DF' is not defined
# Google and chatgpt around the same, told me to define DF

NameError: name 'DF' is not defined

In [27]:
# 8.4D
df.describe(
# SyntaxError: incomplete input
# around the same, both told me to check the syntax

SyntaxError: incomplete input (2300546251.py, line 3)

In [29]:
# 8.4E
df.group_by("col1")["col2"].describe()
#AttributeError: 'DataFrame' object has no attribute 'group_by'


AttributeError: 'DataFrame' object has no attribute 'group_by'

In [31]:
df.groupby("gender")["species"].describle()
# AttributeError: 'SeriesGroupBy' object has no attribute 'describle'
# Around the same for both errors

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [32]:
# 8.4F
df.groupby("genders")["species"].describe()
# KeyError: 'genders'
# The same, both of them had incorrect column name in first response

KeyError: 'genders'

In [34]:
df.groupby("gender")["specie"].describe()
# KeyError: 'Column not found: specie'
# Both were good for this

KeyError: 'Column not found: specie'

In [33]:
# 8.4G
ti.groupby(sex)["age"].describe()
# both were not very good, they did not know that it the input wasn't a string

NameError: name 'sex' is not defined

In [21]:
# Question 9
# Yes

# Chatbot summary


Here's a summary of our exchanges:

Summary:
1. Difference Between Attributes and Methods:

Attribute: A property of an object accessed without parentheses (e.g., df.shape), representing characteristics or properties of the object.

Method: A function associated with an object accessed with parentheses (e.g., df.describe()), performing operations or computations on the object’s data.

2. Use Case for df.dropna() vs. del df['col']:

df.dropna(): Preferred for removing rows with missing values in specific columns, while preserving other columns.

del df['col']: Used for completely removing a column from the DataFrame, which might result in losing valuable data.

3. Order of Operations with del and dropna():

Removing irrelevant columns using del before applying dropna() helps in streamlining data cleaning, making dropna() more focused and efficient.

4. Understanding df.groupby("col1")["col2"].describe():

Groups the DataFrame by col1, then provides descriptive statistics for col2 within each group. This method helps analyze data completeness and distribution within categories.

5. Errors Explanation:

NameError: name 'df' is not defined: Indicates that df has not been defined or initialized.

HTTPError: HTTP Error 404: Not Found: Indicates that the requested resource could not be found on the server.

FileNotFoundError: [Errno 2] No such file or directory: 'villager.csv': Indicates that the specified file could not be found.

SyntaxError: incomplete input: Indicates that the code is incomplete and cannot be parsed.

AttributeError: 'SeriesGroupBy' object has no attribute 'describle': Indicates a typo in method name; should be describe() instead of describle.

KeyError: 'genders' in groupby: Indicates that the column 'genders' does not exist in the DataFrame.

NameError: name 'sex' is not defined: Indicates that the variable or column 'sex' has not been defined or initialized.

This summary captures the main points and explanations provided during our discussion, relevant for your homework assignment.
